### PR TITLE
Arrival Endpoints & adjustments

### PR DESCRIPTION
**Pull request by tifapp**

### CODE REVIEW

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  conn: SQLExecutable,
+  userId: string,
+  input: SetArrivalStatusInput
+) =>
+  conn
+    .queryResults<AttendanceStatus>(
+      `
+        SELECT e.id, e.latitude, e.longitude,
+        CASE
+          WHEN e.latitude != ? OR e.longitude != ? THEN 'invalid'
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: Use table associating userid with desired location to attend. Insert or delete records accordingly from that table.

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          WHEN TIMESTAMPDIFF(HOUR, NOW(), e.startTime) > ? THEN 'early'
+          WHEN TIMESTAMPDIFF(HOUR, NOW(), e.startTime) <= ? THEN 'on-time'
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: Just "success", move "early" or "on-time" to the details query

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+) => {
+  router.postWithValidation(
+    "/arrived",
+    { bodySchema: SetArrivalStatusSchema },
+    (req, res) =>
+      deleteOldArrivals(conn, res.locals.selfId)
+        .flatMapSuccess(() => countExistingArrivals(conn, res.locals.selfId, environment.maxArrivals))
+        .flatMapSuccess(() => getAttendingEvents(conn, res.locals.selfId, req.body.location.latitude, req.body.location.longitude, req.body.events, environment.eventStartWindowInHours))
+        .flatMapSuccess(attendanceStatuses =>
+          (req.body.persistArrivalStatus
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: default save the attendance status

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    ).mapSuccess(() => "removed-arrival" as const)
+
+export const setArrivalStatusRouter = (
+  environment: ServerEnvironment,
+  router: ValidatedRouter
+) => {
+  router.postWithValidation(
+    "/arrived",
+    { bodySchema: SetArrivalStatusSchema },
+    (req, res) =>
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: possibly use inside a transaction

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  conn
+    .queryResults( // TO DECIDE: if event length limit or limit in how far in advance event can be scheduled, then we can also delete outdated arrivals
+      `
+        DELETE FROM userArrivals
+        WHERE userId = :userId
+        AND (
+            ST_Distance_Sphere(
+                POINT(latitude, longitude),
+                POINT(:latitude, :longitude)
+            ) > 1000
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: load distance value dynamically?

Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: distance varying between events? add to event details

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+export const countExistingArrivals = (
+  conn: SQLExecutable,
+  userId: string,
+  arrivalsLimit: number
+) =>
+  conn
+    .queryResults<number>(
+      `
+        SELECT COUNT(*) FROM userArrivals WHERE userId = :userId;
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: Check if possible in a checkconstraint

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  longitude: number,
+  events: number[],
+  hoursBeforeEventStart: number
+) =>
+  conn
+    .queryResults<AttendanceStatus>(
+      `
+        SELECT e.id, e.latitude, e.longitude,
+        CASE
+          WHEN e.latitude != :latitude OR e.longitude != :longitude THEN 'invalid'
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: send invalid if moved/ended, success for other cases

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+export const insertArrival = (
+  conn: SQLExecutable,
+  userId: string,
+  latitude: number,
+  longitude: number
+) =>
+  conn
+    .queryResult(
+      `
+        INSERT IGNORE INTO userArrivals (userId, latitude, longitude) VALUES (:userId, :latitude, :longitude);
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: Check if delete-sert exists

File: APILambda/events/arrivals/setArrivalStatus.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+              res.locals.selfId,
+              req.body.location.latitude,
+              req.body.location.longitude
+            )
+            : deleteArrival(
+              conn,
+              res.locals.selfId,
+              req.body.location.latitude,
+              req.body.location.longitude
+            )).mapSuccess((result) => result === "removed-arrival" ? ({ status: 201, attendanceStatuses: null }) : ({ status: result === "existing-arrival" ? 201 : 200, attendanceStatuses })))
```
Comment (tifapp) [inline APILambda/events/arrivals/setArrivalStatus.ts]: Separate out when adding notifications/chat

File: APILambda/events/arrivals/getUpcomingEvents.test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+describe("getUpcomingEvents tests", () => {
+  resetDatabaseBeforeEach()
+
+  it("should return 204 with an empty array if the user has no upcoming events", async () => {
+    const attendeeToken = await createUserAndUpdateAuth(
+      global.defaultUser
+    )
+
+    expect(withEmptyResponseBody(await callGetUpcomingEvents(attendeeToken))).toMatchObject({
+      status: 204,
```
Comment (seanim0920) [inline APILambda/events/arrivals/getUpcomingEvents.test.ts:19]: use empty array instead

File: APILambda/events/arrivals/getUpcomingEvents.test.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+    const { value: { insertId: arrivedTestEventId } } = await createEvent(
+      conn,
+      { ...testEvents[0], ...eventLocation, startTimestamp: new Date(new Date().setHours(new Date().getHours() + 12)), endTimestamp: new Date(new Date().setFullYear(new Date().getFullYear() + 1)) },
+      randomUUID()
+    )
+
+    const { value: { insertId: ongoingTestEventId } } = await createEvent(
+      conn,
+      { ...testEvents[0], ...eventLocation, startTimestamp: new Date(new Date().setHours(new Date().getHours() - 12)), endTimestamp: new Date(new Date().setFullYear(new Date().getFullYear() + 1)) },
```
Comment (seanim0920) [inline APILambda/events/arrivals/getUpcomingEvents.test.ts:49]: use utils to clean up current date format

File: APILambda/events/arrivals/getUpcomingEvents.test.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  it("should return 200 with an array of events if the user has upcoming events", async () => {
+    const attendeeToken = await createUserAndUpdateAuth(
+      global.defaultUser
+    )
+
+    const eventLocation = { latitude: 50, longitude: 50 }
+
+    await callSetArrival(attendeeToken, {
+      location: eventLocation
```
Comment (seanim0920) [inline APILambda/events/arrivals/getUpcomingEvents.test.ts:32]: add arrivalMetersRadius

File: APILambda/events/arrivals/getUpcomingEvents.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+  eventIds: number[];
+  location: {
+    latitude: number;
+    longitude: number;
+  };
+  arrivalRadiusMeters: number;
+  isArrived: boolean;
+}
+
+const mapEventsToRegions = (events: DatabaseEvent[]): EventRegion[] => {
```
Comment (seanim0920) [inline APILambda/events/arrivals/getUpcomingEvents.ts:16]: Add test for this

File: APILambda/events/arrivals/setArrivalStatus.test.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+    const eventResp = await callCreateEvent(attendeeToken, { ...testEvents[0], ...eventLocation })
+
+    await callSetArrival(attendeeToken, {
+      location: eventLocation
+    })
+
+    expect(await callGetEvent(attendeeToken, parseInt(eventResp.body.id))).toMatchObject({
+      body: {
+        arrivalStatus: "arrived"
```
Comment (seanim0920) [inline APILambda/events/arrivals/setArrivalStatus.test.ts:60]: arrivalStatus -> isArrived

File: APILambda/events/getEventById.ts [hunk 15]
```diff
@@ -24,14 +40,14 @@ export const getEventByIdRouter = (
   router: ValidatedRouter
 ) => {
   router.getWithValidation(
-    "/:eventId",
```
Comment (seanim0920) [inline APILambda/events/getEventById.ts:27]: try moving /upcoming before /:eventId router

File: APILambda/shared/SQL.ts [hunk 16]
```diff
@@ -12,6 +12,7 @@ export type DatabaseEvent = {
   longitude: number
   shouldHideAfterStartDate: boolean
   isChatEnabled: boolean
+  arrivalStatus: string
```
Comment (seanim0920) [inline APILambda/shared/SQL.ts:15]: isArrived

File: APILambda/test/testApp.ts [hunk 17]
```diff
@@ -4,7 +4,9 @@ import { ServerEnvironment } from "../env.js"
 
 export const testEnv: ServerEnvironment = {
   // use env vars
-  environment: process.env.TEST_ENV === "staging" ? "staging" : "dev"
+  environment: process.env.TEST_ENV === "staging" ? "staging" : "dev",
+  eventStartWindowInHours: 1,
+  maxArrivals: 4
```
Comment (seanim0920) [inline APILambda/test/testApp.ts:9]: Can be modified for individual test

**Review by suryatiru — APPROVED** [review]

**Pull request merged**